src/facade/dragonfly_connection.cc

```diff
@@ -2254,4 +2288,173 @@ void ResetStats() {
     io_req_size_hist->Clear();
 }
 
+bool Connection::OffloadBackpressureToDiskIfNeeded(absl::FunctionRef<MessageHandle()> handle) {
+  // Offload only when dispatch_q_ crosses watermark or when backing queue already
+  // has pending items.
+  if (backing_queue_ &&
+      ((dispatch_q_.size() > backing_queue_->Watermark()) || !backing_queue_->Empty())) {
```

Comment:
i prefer that `DiskOffloadThreshold()` would be in QueueBackpressure rather than in `backing_queue_` , this way you do not need to touch backing_queue_ object during the happy path.